### PR TITLE
tests(*) 03-cis-sec.test.sh

### PR DESCRIPTION
**Pull request by kikito**

### CODE REVIEW

File: tests/03-cis-sec.test.sh [hunk 1]
```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env bash
+
+function run_test {
+  # the suite name below will only be used when rtunning this file directly, when
+  # running through "test.sh" it must be provided using the "--suite" option.
+  tinitialize "CIS-Sec tests" "${BASH_SOURCE[0]}"
+
+  tchapter "test $KONG_DOCKER_TAG"
+  ttest "CIS-Sec for docker-compose"
```
Comment (Tieske) [inline tests/03-cis-sec.test.sh:9]:
```suggestion
  tinitialize "Docker-Kong test suite" "${BASH_SOURCE[0]}"

  tchapter "CIS-Sec tests $KONG_DOCKER_TAG"
  ttest "CIS-Sec for docker-compose"
```

File: tests/03-cis-sec.test.sh [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    # Despite the README implying that they do.
+    # Cloning the repo locally and running the .sh file seems to work.
+    # Probably it needs some dev dependencies that my local machine already has
+
+    # Note that the tests assume that your `sed` is GNU-sed instead of the default MacOS'
+    # `brew install gnu-sed` + set the PATH as explained in `brew info gnu-sed`
+
+    if ! git clone https://github.com/docker/docker-bench-security.git tests/docker-bench-security; then
+      echo -e "$T_COLOR_RED  Could not clone docker-bench-security $T_COLOR_CLEAR"
+      tfailure
```
Comment (Tieske) [inline tests/03-cis-sec.test.sh:35]:
```suggestion
      tmessage "Could not clone docker-bench-security"
      tfailure
```

File: tests/03-cis-sec.test.sh [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+      # * 5.1 is AppArmor, which is not available on MacOS
+      # * 5.12 is "root fs is mounted as readonly". The tmpfs option isn't supported in docker for mac
+      MAC_EXCLUDE_TESTS=5_1,5_12
+
+      pushd tests/docker-bench-security
+      if ./docker-bench-security.sh -i kong -e $MAC_EXCLUDE_TESTS &&
+         [ -f "docker-bench-security.sh.log" ]
+      then
+        echo -e "$T_COLOR_RED  Could not execute docker-bench-security $T_COLOR_CLEAR"
+        tfailure
```
Comment (Tieske) [inline tests/03-cis-sec.test.sh:47]:
```suggestion
        tmessage "Could not execute docker-bench-security"
        tfailure
```

File: tests/03-cis-sec.test.sh [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      -v /usr/lib/systemd:/usr/lib/systemd:ro \
+      -v /var/lib:/var/lib:ro \
+      -v /var/run/docker.sock:/var/run/docker.sock:ro \
+      --label docker_bench_security \
+      docker/docker-bench-security
+  fi
+
+  if cat docker-bench-security.sh.log | grep WARN | grep kong -B 1; then
+    echo -e "$T_COLOR_RED  Found warnings in docker-bench-security report $T_COLOR_CLEAR"
+    tfailure
```
Comment (Tieske) [inline tests/03-cis-sec.test.sh:84]:
```suggestion
    tmessage "Found warnings in docker-bench-security report"
    tfailure
```

**Review by Tieske — CHANGES_REQUESTED** [review]

**Review by dndx — APPROVED** [review]

**Review by Tieske — APPROVED** [review]
lgtm

**Comment by murillopaula** [comment]
sorry haven't had bandwidth to review this one... can you take a look at the failing tests? also there are some pending FIXME comments, should we merge with those?

**Comment by Tieske** [comment]
hmmm I assumed it to be the docker rate limit kicking in, but it seems all jobs fail on the same test. So yes, we need the tests fixed. Thx @murillopaula

File: compose/docker-compose.yml [hunk 5]
```diff
@@ -70,20 +70,54 @@ services:
...
       interval: 10s
       timeout: 10s
       retries: 10
-    restart: on-failure
+    restart: on-failure:5
+
+
+    # Hardcoded because Docker-compose cant't handle variable interpolations for integer values
+    # see https://github.com/docker/compose/issues/8297
+    cpu_shares: 1024
```
Comment (hutchic) [inline compose/docker-compose.yml:94]:
```suggestion
```

This is applicable to docker-compose version 2 only and is the current build failure

```
Server: openresty

Compose file contains unsupported options:

cpu_shares: Set resource limits using deploy.resources
```

File: compose/docker-compose.yml [hunk 6]
```diff
@@ -70,20 +70,54 @@ services:
...
+          memory: ${KONG_MEMORY_LIMIT:-2g}
+
+    security_opt:
+      # FIXME: check that apparmor setting works correctly on Linux (can't test on MacOS)
+      - "apparmor:docker-default"
+      - no-new-privileges
+
+    # Hardcoded because Docker-compose cant't handle variable interpolations for integer values
+    # see https://github.com/docker/compose/issues/8297
+    pids_limit: 100
```
Comment (hutchic) [inline compose/docker-compose.yml:119]:
```suggestion
```

docker-compose version three docs say "this hasn't been introduced"

> pids_limit: This option has not been introduced in version: "3.x" Compose files.

**Comment by hutchic** [comment]
after the two docker-compose version three compatibility fixes I'm hitting the following locally
```
kong_kong.1.9v8lhvns3xsp@ip-172-31-30-251    | stack traceback:
kong_kong.1.9v8lhvns3xsp@ip-172-31-30-251    |  [C]: in function 'assert'
kong_kong.1.9v8lhvns3xsp@ip-172-31-30-251    |  /usr/local/share/lua/5.1/kong/init.lua:448: in function 'init'
kong_kong.1.9v8lhvns3xsp@ip-172-31-30-251    |  init_by_lua:3: in main chunk
kong_kong.1.pkwhtdjok3c4@ip-172-31-30-251    | nginx: [warn] the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
kong_kong.1.pkwhtdjok3c4@ip-172-31-30-251    | 2021/05/05 20:20:46 [error] 1#0: init_by_lua error: /usr/local/share/lua/5.1/kong/init.lua:448: [PostgreSQL error] failed to retrieve PostgreSQL server_version_num: host or service not provided, or not known
```

**Review by gszr — APPROVED** [review]

**Pull request merged**